swarm/storage/pushsync/pusher_test.go

```diff
@@ -0,0 +1,414 @@
+}
+
+// loopback implements PubSub as a central subscription engine,
+// ie a msg sent is received by all handlers registered for the topic
+type loopBack struct {
+	async    bool
+	addr     []byte
+	handlers map[string][]func(msg []byte, p *p2p.Peer) error
+}
+
+func newLoopBack(async bool) *loopBack {
+	return &loopBack{
+		async:    async,
+		addr:     make([]byte, 32),
+		handlers: make(map[string][]func(msg []byte, p *p2p.Peer) error),
+	}
+}
+
+// Register subscribes to a topic with a handler
+func (lb *loopBack) Register(topic string, _ bool, handler func(msg []byte, p *p2p.Peer) error) func() {
+	lb.handlers[topic] = append(lb.handlers[topic], handler)
+	return func() {}
+}
+
+// Send publishes a msg with a topic and directly calls registered handlers with
+// that topic
+func (lb *loopBack) Send(to []byte, topic string, msg []byte) error {
+	if lb.async {
+		go lb.send(to, topic, msg)
+		return nil
+	}
+	return lb.send(to, topic, msg)
+}
+
+func (lb *loopBack) send(to []byte, topic string, msg []byte) error {
+	p := p2p.NewPeer(enode.ID{}, "", nil)
+	for _, handler := range lb.handlers[topic] {
+		if err := handler(msg, p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BaseAddr needed to implement PubSub interface
+func (lb *loopBack) BaseAddr() []byte {
+	return lb.addr
+}
+
+// testPushSyncIndex mocks localstore and provides subscription and setting synced status
+// it implements PushSyncIndex interface
+type testPushSyncIndex struct {
+	i, total int
+	tagIDs   []uint32 //
+	tags     *chunk.Tags
+	sent     *sync.Map // to store time of send for retry
+	synced   chan int  // to check if right amount of chunks
+}
+
+func newTestPushSyncIndex(chunkCnt int, tagIDs []uint32, tags *chunk.Tags, sent *sync.Map) *testPushSyncIndex {
+	return &testPushSyncIndex{
+		i:      0,
+		total:  chunkCnt,
+		tagIDs: tagIDs,
+		tags:   tags,
+		sent:   sent,
+		synced: make(chan int),
+	}
+}
+
+// SubscribePush allows iteration on the hashes and mocks the behaviour of localstore
+// push index
+// we keep track of an index so that each call to SubscribePush knows where to start
+// generating the new fake hashes
+// Before the new fake hashes it iterates over hashes not synced yet
+func (t *testPushSyncIndex) SubscribePush(context.Context) (<-chan storage.Chunk, func()) {
+	chunks := make(chan storage.Chunk)
+	tagCnt := len(t.tagIDs)
+	quit := make(chan struct{})
+	stop := func() { close(quit) }
+	go func() {
+		// feed fake chunks into the db, hashes encode the order so that
+		// it can be traced
+		feed := func(i int) bool {
+			// generate fake hashes that encode the chunk order
+			addr := make([]byte, 32)
+			binary.BigEndian.PutUint64(addr, uint64(i))
+			// remember when the chunk was put
+			// if sent again, dont modify the time
+			t.sent.Store(i, time.Now())
+			// increment stored count on tag
+			tagID := t.tagIDs[i%tagCnt]
+			if tag, _ := t.tags.Get(tagID); tag != nil {
```

Comment:
the tag should  exist. fail if it is not found

Reply:
no. i deliberately keep a nonexistent tag to test

Reply:
...not sure if i get the intent